### ISSUE TITLE
Fill out unimplemented expression cases in the ASTPrinter

### ISSUE THREAD

**Issue by hamishknight** [apple/swift#61601]
In #40691 we added initial support for printing `Expr` nodes in the ASTPrinter. However there are still some expressions left unimplemented, including:

- [ ] TenaryExpr
- [ ] IsExpr
- [ ] TapExpr
- [ ] TypeJoinExpr
- [ ] CoerceExpr
- [ ] OneWayExpr
- [ ] ClosureExpr
- [ ] KeyPathExpr
- [ ] SequenceExpr
- [ ] SuperRefExpr
- [ ] SubscriptExpr
- [ ] EnumIsCaseExpr
- [ ] ForceValueExpr
- [ ] KeyPathDotExpr
- [ ] CaptureListExpr
- [ ] DynamicTypeExpr
- [ ] OptionalTryExpr
- [ ] PrefixUnaryExpr
- [ ] BindOptionalExpr
- [ ] ObjCSelectorExpr
- [ ] PostfixUnaryExpr
- [ ] TupleElementExpr
- [ ] ObjectLiteralExpr
- [ ] DynamicMemberRefExpr
- [ ] DynamicSubscriptExpr
- [ ] UnresolvedMemberExpr
- [ ] EditorPlaceholderExpr
- [ ] OverloadedDeclRefExpr
- [ ] UnresolvedDeclRefExpr
- [ ] UnresolvedPatternExpr
- [ ] KeyPathApplicationExpr
- [ ] DotSyntaxBaseIgnoredExpr
- [ ] MagicIdentifierLiteralExpr
- [ ] OtherConstructorDeclRefExpr
- [ ] RebindSelfInConstructorExpr
- [ ] MakeTemporarilyEscapableExpr
- [ ] InterpolatedStringLiteralExpr

We should fill in the implementation for these cases.

**Comment by hamishknight** [comment]
cc @louisdh

**Comment by jakedves** [comment]
Hey, I'm interested in looking at this but I'm not super familiar with the codebase. Do you think this could be an appropriate StarterBug? Thanks

**Comment by hamishknight** [comment]
Sure, I think this would make a suitable StarterBug, it should give you a good sense of how the AST is structured. I'm happy to be pinged with questions :)

Feel free to pick off individual nodes, though communicate here what nodes you want to take so people don't end up working on the same thing.

**Comment by jakedves** [comment]
Sure, thanks. I'll ping with any questions I get, and I'll start with IsExpr and ForceValueExpr.

**Comment by XaurDesu** [comment]
Is there room for another contributor? I could take a look at this if you guys don't mind.

**Comment by xedin** [comment]
I don’t think anybody is working on this at the moment, so feel free to take it!

**Comment by XaurDesu** [comment]
Alright! i'm taking a look into it this weekend. I'm currently a college student and i'm doing my finals, but once i'm done with those i will be making a pull request. Might read a bit of the codebase in the meantime

**Comment by XaurDesu** [comment]
Hey guys, given that there are quite a few things, do you think it would be more sensible to try and make a pull request when one is finished rather than make a big one with all of them at the same time? That way we can start merging the stuff just as i finish it, instead of waiting for the full ticket to be completed, and we can keep track of them through the checklist in this issue.

**Comment by xedin** [comment]
Either way is fine, if you want to do a PR per, go for it.

**Comment by Rajveer100** [comment]
All changes I suppose would be done in the `ASTPrinter.cpp` file?

**Comment by Rajveer100** [comment]
@hamishknight Could you give a brief info about this issue, although I have gone through the previous pull request and learning about AST.

**Comment by hamishknight** [comment]
> All changes I suppose would be done in the ASTPrinter.cpp file?

Yes, that's correct, specifically the `ASTPrinter::visitXXXExpr` methods that are currently empty.


> Could you give a brief info about this issue, although I have gone through the previous pull request and started digging about AST.

@Rajveer100 This is about being able to print back out expressions from their AST node, e.g we should be able to print `<expr> as <Type>` given the `CoerceExpr` that describes it (where `<expr>` and `<Type>` are replaced with the expression and type used).

**Comment by Rajveer100** [comment]
> @Rajveer100 This is about being able to print back out expressions from their AST node, e.g we should be able to print `<expr> as <Type>` given the `CoerceExpr` that describes it (where `<expr>` and `<Type>` are replaced with the expression and type used).

You mean to say, if the Abstract Syntax Tree (AST) had to evaluate an expression, like a simple math equation (3 * 7 + 13), the resulting type would be of some type (ex. `<expr>`) and the 'ASTPrinter' would be able to print it in a certain form (ex. `<Type>`)?

**Comment by Rajveer100** [comment]
Could you also give some example code snippet to 'test' any suitable cases (not necessarily among the implemented cases), for more familiarity.

**Comment by hamishknight** [comment]
@Rajveer100 You can try it out with expressions like `3 * 7 + 13` with `./swift -frontend -print-ast /path/to/file.swift`. It should just print back out the original expression.

**Comment by Rajveer100** [comment]
Well actually I use Xcode WorkSpace, in the scheme, just after the `-typecheck` I added the `-print-ast` flag and it gave the following output:

```
/swift-project/build/CodeBase/DebugArea.swift:29:7: warning: result of operator '+' is unused
3 * 7 + 13
~~~~~ ^ ~~

3 * 7 + 13
```

Without the flag:

```
/swift-project/build/CodeBase/DebugArea.swift:29:7: warning: result of operator '+' is unused
3 * 7 + 13
~~~~~ ^ ~~
```

(I suppose that was the intended result?)

And I didn't use `-frontend` flag as it threw 'unknown argument' error.

**Comment by jacbro2021** [comment]
Is work still needed for this? I am looking to pick up a first issue but I see there was already a PR merged.

**Comment by vanshkamra12** [comment]
@AnthonyLatsis @hamishknight can i work on this ...?????

**Comment by Anandb71** [comment]
I am taking this on.
